File: src/main/java/org/forweb/soldiers/service/GeometryService.java
package org.forweb.soldiers.service;

import org.forweb.soldiers.controller.PersonWebSocketEndpoint;
import org.forweb.soldiers.entity.zone.AbstractZone;
import org.forweb.soldiers.utils.shapes.Line;
import org.forweb.soldiers.utils.shapes.Point;

import java.awt.Rectangle;

public class GeometryService {

    static Rectangle getRectangle(AbstractZone zone) {
        return new Rectangle(zone.getX(), zone.getY(), zone.getWidth(), zone.getHeight());
    }


    public static Point[] getIntersectionPoint(Line projectileTrace, Rectangle rectangle) {

        Point[] p = new Point[4];
        // Top line
        p[0] = GeometryService.getIntersectionPoint(
                projectileTrace,
                new Line(
                        new Point(rectangle.getX(), rectangle.getY()),
                        new Point(rectangle.getX() + rectangle.getWidth(), rectangle.getY())
                )
        );
        // Right side
        p[1] = GeometryService.getIntersectionPoint(
                projectileTrace,
                new Line(
                        new Point(rectangle.getX() + rectangle.getWidth(), rectangle.getY()),
                        new Point(rectangle.getX() + rectangle.getWidth(), rectangle.getY() + rectangle.getHeight())
                )
        );
        // Bottom line
        p[2] = GeometryService.getIntersectionPoint(
                projectileTrace,
                new Line(
                        new Point(rectangle.getX(), rectangle.getY() + rectangle.getHeight()),
                        new Point(rectangle.getX() + rectangle.getWidth(), rectangle.getY() + rectangle.getHeight())
                )
        );
        // Left side...
        p[3] = GeometryService.getIntersectionPoint(
                projectileTrace,
                new Line(
                        new Point(rectangle.getX(), rectangle.getY()),
                        new Point(rectangle.getX(), rectangle.getY() + rectangle.getHeight())
                )
        );
        return p;
    }

    private static Point getIntersectionPoint(Line line1, Line line2) {
        Point a1 = line1.getA();
        Point b1 = line1.getB();
        Point a2 = line2.getA();
        Point b2 = line2.getB();

        double x1 = a1.getX();
        double y1 = a1.getY();
        double x2 = b1.getX();
        double y2 = b1.getY();
        double x3 = a2.getX();
        double y3 = a2.getY();
        double x4 = b2.getX();
        double y4 = b2.getY();

        double v1 = (x4 - x3) * (y1 - y3) - (y4 - y3) * (x1 - x3);
        double v2 = (x4 - x3) * (y2 - y3) - (y4 - y3) * (x2 - x3);
        double v3 = (x2 - x1) * (y3 - y1) - (y2 - y1) * (x3 - x1);
        double v4 = (x2 - x1) * (y4 - y1) - (y2 - y1) * (x4 - x1);
        if (!((v1 * v2 <= 0) && (v3 * v4 <= 0))) {
            return null;
        }

        Point p = null;
        double d = (x1 - x2) * (y3 - y4) - (y1 - y2) * (x3 - x4);
        if (d != 0) {
            double xi = ((x3 - x4) * (x1 * y2 - y1 * x2) - (x1 - x2) * (x3 * y4 - y3 * x4)) / d;
            double yi = ((y3 - y4) * (x1 * y2 - y1 * x2) - (y1 - y2) * (x3 * y4 - y3 * x4)) / d;

            p = new Point(xi, yi);
        }
        return p;
    }


    public static Point[] getCircleLineIntersectionPoint(Point linePointA, Point linePointB, Point personCenter) {
        double baX = linePointB.getX() - linePointA.getX();
        double baY = linePointB.getY() - linePointA.getY();
        double caX = personCenter.getX() - linePointA.getX();
        double caY = personCenter.getY() - linePointA.getY();

        double a = baX * baX + baY * baY;
        double bBy2 = baX * caX + baY * caY;
        double c = caX * caX + caY * caY - (double) PersonWebSocketEndpoint.PERSON_RADIUS * (double) PersonWebSocketEndpoint.PERSON_RADIUS;

        double pBy2 = bBy2 / a;
        double q = c / a;

        double disc = pBy2 * pBy2 - q;
        if (disc < 0) {
            return null;
        }
        double tmpSqrt = Math.sqrt(disc);
        double abScalingFactor1 = -pBy2 + tmpSqrt;
        double abScalingFactor2 = -pBy2 - tmpSqrt;

        Point p1 = new Point(linePointA.getX() - baX * abScalingFactor1, linePointA.getY()
                - baY * abScalingFactor1);
        if (disc == 0) {
            return new Point[]{p1};
        }
        Point p2 = new Point(linePointA.getX() - baX * abScalingFactor2, linePointA.getY()
                - baY * abScalingFactor2);
        return new Point[]{p1, p2};
    }
}

File: src/main/webapp/js/utils/GeometryService.js
var GeometryService = {
    lineIntersectCircle: function(line, circle) {
        var pointB = line.b;
        var pointA = line.a;
        if(pointA.x == pointB.x && pointA.y == pointB.y) {
            return Math.pow(pointA.x - circle.x, 2) + Math.pow(pointA.y - circle.y, 2) == circle.radius * circle.radius?[pointA]:[];
        } else {
            var baX = pointB.x - pointA.x;
            var baY = pointB.y - pointA.y;
            var caX = circle.x - pointA.x;
            var caY = circle.y - pointA.y;
            var a = baX * baX + baY * baY;
            var bBy2 = baX * caX + baY * caY;
            var c = caX * caX + caY * caY - circle.radius * circle.radius;
            var pBy2 = bBy2 / a;
            var q = c / a;
            var disc = pBy2 * pBy2 - q;
            if(disc <= 1.0E-4 && disc >= -1.0E-4) {
                disc = 0;
            }

            if(disc < 0) {
                return [];
            } else {
                var tmpSqrt = Math.sqrt(disc);
                var abScalingFactor1 = -pBy2 + tmpSqrt;
                var abScalingFactor2 = -pBy2 - tmpSqrt;
                var p1 = {x: pointA.x - baX * abScalingFactor1, y: pointA.y - baY * abScalingFactor1};
                var out1 = GeometryService.getMiddle(pointA, pointB, p1);
                if(disc == 0) {
                    return out1 == p1?[p1]:[];
                } else {
                    var p2 = {x: pointA.x - baX * abScalingFactor2, y: pointA.y - baY * abScalingFactor2};
                    var out2 = GeometryService.getMiddle(pointA, pointB, p2);
                    return out1 == p1?(out2 == p2?[p1, p2]:[p1]):(out2 == p2?[p2]:[]);
                }
            }
        }
    },
    getMiddle: function(pointA, pointB, pointC) {
        var ab = GeometryService.getDistance(pointA, pointB);
        var bc = GeometryService.getDistance(pointB, pointC);
        var ca = GeometryService.getDistance(pointC, pointA);
        return ab >= bc && ab >= ca?pointC:(ca >= ab && ca >= bc?pointB:pointA);
    },
    getDistance: function(point1, point2) {
        var x1 = point1.x;
        var y1 = point1.y;
        var x2 = point2.x;
        var y2 = point2.y;
        return Math.sqrt(Math.pow(x2 - x1, 2.0) + Math.pow(y2 - y1, 2.0));
    }
};